keps/sig-node/3063-dynamic-resource-allocation/README.md

```diff
@@ -0,0 +1,2164 @@
+
+* Provide an abstraction layer for resource requests, i.e., something like a
+  “I want some kind of GPU”. Users will need to know about specific
+  resource drivers and which parameters they support. Portability of
+  workloads could be added on top of this proposal by introducing the
+  selection of a resource implementation through labels and
+  standardizing those labels and the associated parameters. The
+  [Resource Class
+  Proposal](https://docs.google.com/document/d/1qKiIVs9AMh2Ua5thhtvWqOqW0MSle_RV3lfriO1Aj6U/edit#heading=h.jzfmfdca34kj)
+  included such an approach.
+
+
+
+## Proposal
+
+The proposal is that a resource driver handles all operations that are specific
+to the resources managed by that driver. This includes operations at
+the control plane level (tracking where in the cluster resources are
+available, helping with pod scheduling decisions, allocating resources
+when requested) as well as the node level (preparing container
+startup). Such a driver can be implemented in arbitrary programming
+languages as long as it supports the resource allocation protocol and
+gRPC interfaces defined in this KEP. Deploying it will not depend on
+reconfiguring core Kubernetes components like the scheduler.
+
+New API objects define parameters for a resource request ("ResourceClaim" in
+the API) and track the state of such a request. The pod spec gets extended to
+reference one or more resource requests. A pod only gets scheduled onto a node
+when all of its requests are reserved for the pod and available on the
+node. This prevents scheduling a pod that then gets stuck on a node while
+waiting for resources to become available.
+
+### User Stories
+
+#### Cluster add-on development
+
+As a hardware vendor, I want to make my hardware available also to applications
+that run in a container under Kubernetes. I want to make it easy for a cluster
+administrator to configure a cluster where some nodes have this hardware.
+
+I develop two components, one that runs as part of the Kubernetes control plane
+and one that runs on each node, and package those inside container images. YAML
+files describe how to deploy my software on a Kubernetes cluster that supports
+dynamic resource allocation.
+
+Documentation for administrators explains how the nodes need to be set
+up. Documentation for users explains which parameters control the behavior of
+my hardware and how to use it inside a container.
+
+#### Cluster configuration
+
+As a cluster administrator, I want to make GPUs from vendor ACME available to users
+of that cluster. I prepare the nodes and deploy the vendor's components with
+`kubectl create`.
+
+I create a ResourceClass for the hardware with parameters that only I as the
+administrator am allowed to choose, like for example running a command with
+root privileges that does some cluster-specific initialization for each allocation:
+```
+apiVersion: gpu.acme.com/v1
+kind: GPUInit
+metadata:
+  name: acme-gpu-init
+initCommand:
```

Comment:
The "terrifying" was intentional here, to show that this is still somewhat safe because such an initCommand is only valid for a ResourceClass and thus usable by an admin. I added a "DANGER!" comment inline to make it even more obvious.

I can also replace it, but then the difference between user- and admin-supplied parameters becomes less clear. Shall I do it nonetheless?